docker/Dockerfile

```diff
@@ -1,7 +1,24 @@
 ARG PYTHON_VERSION=2
-FROM docker.io/python:${PYTHON_VERSION}-alpine3.8
 
-COPY . /synapse
+###
+### Stage 0: builder
+###
+FROM docker.io/python:${PYTHON_VERSION}-alpine3.8 as builder
+
+# first of all, install the runtime deps; this layer can then be used in the
+# final stage as well.
+
+RUN apk add --no-cache --virtual .runtime_deps \
+        libffi \
+        libjpeg-turbo \
+        libressl \
+        libxslt \
+        libpq \
+        zlib \
+        su-exec
+
+# install the OS build deps, and build wheels for the python libs which have
+# slow build steps, as more layers which can be cached independently
 
 RUN apk add --no-cache --virtual .build_deps \
```

Comment:
As said above, we don't need .build_deps here. We shouldn't need --no-cache either, because this is a throwaway build container where we don't care about caches making the image larger